File: ontrack-extension-git/src/test/java/net/nemerosa/ontrack/extension/git/GitMockConfig.java
package net.nemerosa.ontrack.extension.git;

import net.nemerosa.ontrack.git.GitRepository;
import net.nemerosa.ontrack.git.GitRepositoryClient;
import net.nemerosa.ontrack.git.GitRepositoryClientFactory;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Primary;
import org.springframework.context.annotation.Profile;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

@Configuration
@Profile(GitMockConfig.PROFILE_GIT_MOCK)
public class GitMockConfig {

    public static final String PROFILE_GIT_MOCK = "git.mock";

    @Bean
    public GitRepositoryClient testGitRepositoryClient() {
        return mock(GitRepositoryClient.class);
    }

    @Bean
    @Primary
    public GitRepositoryClientFactory repositoryClientFactory() {
        GitRepositoryClientFactory factory = mock(GitRepositoryClientFactory.class);
        when(factory.getClient(any(GitRepository.class))).thenReturn(testGitRepositoryClient());
        return factory;
    }

}

File: ontrack-extension-git/src/test/java/net/nemerosa/ontrack/extension/git/GitMockConfig.kt
package net.nemerosa.ontrack.extension.git

import io.mockk.every
import io.mockk.mockk
import net.nemerosa.ontrack.git.GitRepositoryClient
import net.nemerosa.ontrack.git.GitRepositoryClientFactory
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Primary
import org.springframework.context.annotation.Profile

@Configuration
@Profile(GitMockConfig.Companion.PROFILE_GIT_MOCK)
class GitMockConfig {
    @Bean
    fun testGitRepositoryClient(): GitRepositoryClient {
        return mockk(relaxed = true)
    }

    @Bean
    @Primary
    fun repositoryClientFactory(): GitRepositoryClientFactory {
        val factory = mockk<GitRepositoryClientFactory>(relaxed = true)
        every {
            factory.getClient(
                any(),
                any()
            )
        } returns testGitRepositoryClient()
        return factory
    }

    companion object {
        const val PROFILE_GIT_MOCK: String = "git_mock"
    }
}
